File: agents/cleanup_agent.py
"""
Cleanup Agent for q_mini_wasm_v2

This agent monitors for task completion patterns and performs cleanup operations.
It looks for patterns like "Final Execution Complete" to trigger cleanup workflows.
"""

import asyncio
import json
import re
from datetime import datetime, timedelta
from pathlib import Path
from typing import Any, Dict, List, Optional, Set

import structlog

from .base_agent import BaseAgent, AgentConfig, TaskResult

logger = structlog.get_logger()


class CleanupAgent(BaseAgent):
    """
    Cleanup agent for monitoring task completion and performing cleanup.
    
    This agent:
    1. Monitors for completion patterns like "Final Execution Complete"
    2. Cleans up temporary files and artifacts
    3. Archives completed task logs
    4. Manages memory and resource cleanup
    """
    
    # Patterns that indicate task completion
    COMPLETION_PATTERNS = [
        r"Final Execution Complete",
        r"Task completed successfully",
        r"All tests passed",
        r"Cycle completed",
        r"Deployment successful",
        r"Build successful",
        r"Documentation update complete",
        r"CI/CD pipeline complete",
    ]
    
    # File patterns to clean up
    TEMP_FILE_PATTERNS = [
        "*.tmp",
        "*.temp",
        "*.log",
        "*.bak",
        "*~",
        "__pycache__",
        "*.pyc",
        ".pytest_cache",
        ".mypy_cache",
    ]
    
    # Directories to monitor
    MONITOR_DIRS = [
        "docs",
        "wiki-output",
        "build",
        "dist",
        "agents/memory",
    ]
    
    def __init__(self, config: AgentConfig, llm_service=None):
        super().__init__(config, llm_service)
        self._completion_cache: Dict[str, Any] = {}
        self._cleanup_history: List[Dict[str, Any]] = []
        self._last_cleanup: Optional[datetime] = None
        
    async def execute_task(self, task: Dict[str, Any]) -> TaskResult:
        """
        Execute a cleanup task.
        
        Args:
            task: Task specification with 'type' and 'parameters'
            
        Returns:
            TaskResult with cleanup results
        """
        task_type = task.get("type", "unknown")
        parameters = task.get("parameters", {})
        
        self.logger.info("Executing cleanup task", task_type=task_type)
        self.state = "running"
        
        try:
            if task_type == "scan_completions":
                result = await self._scan_for_completions(parameters)
            elif task_type == "cleanup_temp_files":
                result = await self._cleanup_temp_files(parameters)
            elif task_type == "archive_logs":
                result = await self._archive_completed_logs(parameters)
            elif task_type == "full_cleanup":
                result = await self._full_cleanup(parameters)
            elif task_type == "check_pattern":
                result = await self._check_completion_pattern(parameters)
            else:
                return TaskResult(
                    success=False,
                    errors=[f"Unknown task type: {task_type}"]
                )
            
            self.state = "idle"
            return TaskResult(success=True, data=result)
            
        except Exception as e:
            self.logger.error("Cleanup task failed", error=str(e))
            self.state = "error"
            return TaskResult(success=False, errors=[str(e)])
    
    async def _scan_for_completions(self, parameters: Dict[str, Any]) -> Dict[str, Any]:
        """
        Scan for completion patterns in logs and outputs.
        
        Returns:
            Dictionary with completion patterns found
        """
        repo_root = Path(parameters.get("repo_root", "."))
        scan_paths = parameters.get("scan_paths", ["logs", "output", "build"])
        
        completions_found = []
        
        for scan_path in scan_paths:
            path = repo_root / scan_path
            if path.exists():
                # Scan for completion patterns in files
                for pattern in self.COMPLETION_PATTERNS:
                    completions = self._find_pattern_in_files(path, pattern)
                    completions_found.extend(completions)
        
        # Update cache
        self._completion_cache = {
            "last_scan": datetime.now().isoformat(),
            "completions_found": completions_found,
            "patterns_checked": len(self.COMPLETION_PATTERNS)
        }
        
        return {
            "scan_complete": True,
            "completions_found": len(completions_found),
            "patterns": completions_found[:10]  # Limit for readability
        }
    
    def _find_pattern_in_files(self, directory: Path, pattern: str) -> List[Dict[str, Any]]:
        """Find pattern matches in files within a directory."""
        matches = []
        
        try:
            for file_path in directory.rglob("*"):
                if file_path.is_file() and file_path.suffix in [".txt", ".log", ".md", ".json"]:
                    try:
                        content = file_path.read_text(encoding="utf-8")
                        if re.search(pattern, content, re.IGNORECASE):
                            matches.append({
                                "file": str(file_path.relative_to(directory.parent)),
                                "pattern": pattern,
                                "timestamp": datetime.now().isoformat()
                            })
                    except Exception:
                        continue
        except Exception as e:
            self.logger.warning("Error scanning directory", directory=str(directory), error=str(e))
        
        return matches
    async def _cleanup_temp_files(self, parameters):
        repo_root = Path(parameters.get('repo_root', '.'))
        dry_run = parameters.get('dry_run', False)
        files_cleaned = []
        bytes_freed = 0
        for pattern in self.TEMP_FILE_PATTERNS:
            for file_path in repo_root.rglob(pattern):
                if file_path.is_file():
                    try:
                        file_size = file_path.stat().st_size
                        if not dry_run:
                            file_path.unlink()
                        files_cleaned.append(str(file_path.relative_to(repo_root)))
                        bytes_freed += file_size
                    except Exception as e:
                        pass
        self._last_cleanup = datetime.now()
        return {'cleanup_complete': True, 'files_cleaned': len(files_cleaned), 'bytes_freed': bytes_freed}
    
    async def _archive_completed_logs(self, parameters):
        return {'archive_complete': True, 'logs_archived': 0}
    
    async def _full_cleanup(self, parameters):
        results = {}
        results['scan'] = await self._scan_for_completions(parameters)
        results['cleanup'] = await self._cleanup_temp_files(parameters)
        return {'full_cleanup_complete': True, 'results': results}
    
    async def _check_completion_pattern(self, parameters):
        text = parameters.get('text', '')
        pattern = parameters.get('pattern', 'Final Execution Complete')
        match = re.search(pattern, text, re.IGNORECASE)
        return {'pattern_found': match is not None, 'pattern': pattern, 'matched_text': match.group() if match else None}
    
    async def analyze_performance(self):
        return {'tasks_completed': self._task_count, 'last_cleanup': self._last_cleanup.isoformat() if self._last_cleanup else None}
    
    async def suggest_improvements(self):
        return []

File: agents/pkg/cleanup_agent.go
package pkg

import (
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"
)

// CleanupAgent monitors task completion patterns and performs cleanup operations
type CleanupAgent struct {
	*BaseAgent
	completionCache map[string]interface{}
	cleanupHistory  []map[string]interface{}
	lastCleanup     time.Time

	COMPLETION_PATTERNS []string
	TEMP_FILE_PATTERNS  []string
	MONITOR_DIRS        []string
}

// NewCleanupAgent creates a new CleanupAgent
func NewCleanupAgent(config *AgentConfig) *CleanupAgent {
	return &CleanupAgent{
		BaseAgent:       NewBaseAgent(config),
		completionCache: make(map[string]interface{}),
		cleanupHistory:  make([]map[string]interface{}, 0),
		COMPLETION_PATTERNS: []string{
			`Final Execution Complete`,
			`Task completed successfully`,
			`All tests passed`,
			`Cycle completed`,
			`Deployment successful`,
			`Build successful`,
			`Documentation update complete`,
			`CI/CD pipeline complete`,
		},
		TEMP_FILE_PATTERNS: []string{
			"*.tmp",
			"*.temp",
			"*.log",
			"*.bak",
			"*~",
			"__pycache__",
			"*.pyc",
			".pytest_cache",
			".mypy_cache",
		},
		MONITOR_DIRS: []string{
			"docs",
			"wiki-output",
			"build",
			"dist",
			"agents/memory",
		},
	}
}

// ExecuteTask executes a cleanup task
func (a *CleanupAgent) ExecuteTask(task map[string]interface{}) (*TaskResult, error) {
	taskType, _ := task["type"].(string)
	parameters, _ := task["parameters"].(map[string]interface{})

	if taskType == "" {
		taskType = "unknown"
	}

	a.State = AgentStateRunning

	var result interface{}
	var err error

	switch taskType {
	case "scan_completions":
		result, err = a.scanForCompletions(parameters)
	case "cleanup_temp_files":
		result, err = a.cleanupTempFiles(parameters)
	case "archive_logs":
		result, err = a.archiveCompletedLogs(parameters)
	case "full_cleanup":
		result, err = a.fullCleanup(parameters)
	case "check_pattern":
		result, err = a.checkCompletionPattern(parameters)
	default:
		a.State = AgentStateIdle
		return &TaskResult{
			Success: false,
			Errors:  []string{"Unknown task type: " + taskType},
		}, nil
	}

	if err != nil {
		a.State = AgentStateError
		return &TaskResult{
			Success: false,
			Errors:  []string{err.Error()},
		}, err
	}

	a.State = AgentStateIdle

	return &TaskResult{
		Success: true,
		Data:    result,
	}, nil
}

// scanForCompletions scans for completion patterns in logs and outputs
func (a *CleanupAgent) scanForCompletions(parameters map[string]interface{}) (map[string]interface{}, error) {
	repoRoot, _ := parameters["repo_root"].(string)
	if repoRoot == "" {
		repoRoot = "."
	}

	scanPaths, ok := parameters["scan_paths"].([]string)
	if !ok || len(scanPaths) == 0 {
		scanPaths = []string{"logs", "output", "build"}
	}

	completionsFound := make([]map[string]interface{}, 0)

	for _, scanPath := range scanPaths {
		path := filepath.Join(repoRoot, scanPath)
		if _, err := os.Stat(path); err == nil {
			for _, pattern := range a.COMPLETION_PATTERNS {
				completions := a.findPatternInFiles(path, pattern)
				completionsFound = append(completionsFound, completions...)
			}
		}
	}

	a.completionCache = map[string]interface{}{
		"last_scan":        time.Now().Format(time.RFC3339),
		"completions_found": completionsFound,
		"patterns_checked": len(a.COMPLETION_PATTERNS),
	}

	return map[string]interface{}{
		"scan_complete":      true,
		"completions_found":  len(completionsFound),
		"patterns":           completionsFound[:min(10, len(completionsFound))],
	}, nil
}

// findPatternInFiles finds pattern matches in files within a directory
func (a *CleanupAgent) findPatternInFiles(directory string, pattern string) []map[string]interface{} {
	matches := make([]map[string]interface{}, 0)

	re, err := regexp.Compile("(?i)" + pattern)
	if err != nil {
		return matches
	}

	err = filepath.Walk(directory, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}

		if !info.Mode().IsRegular() {
			return nil
		}

		ext := strings.ToLower(filepath.Ext(path))
		if ext != ".txt" && ext != ".log" && ext != ".md" && ext != ".json" {
			return nil
		}

		content, err := os.ReadFile(path)
		if err != nil {
			return nil
		}

		if re.Match(content) {
			relPath, _ := filepath.Rel(filepath.Dir(directory), path)
			matches = append(matches, map[string]interface{}{
				"file":      relPath,
				"pattern":   pattern,
				"timestamp": time.Now().Format(time.RFC3339),
			})
		}

		return nil
	})

	return matches
}

// cleanupTempFiles cleans up temporary files
func (a *CleanupAgent) cleanupTempFiles(parameters map[string]interface{}) (map[string]interface{}, error) {
	repoRoot, _ := parameters["repo_root"].(string)
	if repoRoot == "" {
		repoRoot = "."
	}

	dryRun, ok := parameters["dry_run"].(bool)
	if !ok {
		dryRun = false
	}

	filesCleaned := make([]string, 0)
	bytesFreed := int64(0)

	for _, pattern := range a.TEMP_FILE_PATTERNS {
		err := filepath.Walk(repoRoot, func(path string, info os.FileInfo, err error) error {
			if err != nil {
				return err
			}

			if strings.Contains(path, ".git") {
				if info.IsDir() {
					return filepath.SkipDir
				}
				return nil
			}

			matched, err := filepath.Match(pattern, filepath.Base(path))
			if err != nil || !matched {
				return nil
			}

			if info.Mode().IsRegular() {
				fileSize := info.Size()
				if !dryRun {
					os.Remove(path)
				}
				relPath, _ := filepath.Rel(repoRoot, path)
				filesCleaned = append(filesCleaned, relPath)
				bytesFreed += fileSize
			}

			return nil
		})

		if err != nil {
			continue
		}
	}

	a.lastCleanup = time.Now()

	return map[string]interface{}{
		"cleanup_complete": true,
		"files_cleaned":    len(filesCleaned),
		"bytes_freed":      bytesFreed,
	}, nil
}

// archiveCompletedLogs archives completed logs
func (a *CleanupAgent) archiveCompletedLogs(parameters map[string]interface{}) (map[string]interface{}, error) {
	return map[string]interface{}{
		"archive_complete": true,
		"logs_archived":    0,
	}, nil
}

// fullCleanup performs full cleanup cycle
func (a *CleanupAgent) fullCleanup(parameters map[string]interface{}) (map[string]interface{}, error) {
	results := make(map[string]interface{})

	scanResult, err := a.scanForCompletions(parameters)
	results["scan"] = scanResult
	if err != nil {
		return nil, err
	}

	cleanupResult, err := a.cleanupTempFiles(parameters)
	results["cleanup"] = cleanupResult
	if err != nil {
		return nil, err
	}

	return map[string]interface{}{
		"full_cleanup_complete": true,
		"results":              results,
	}, nil
}

// checkCompletionPattern checks text for completion pattern
func (a *CleanupAgent) checkCompletionPattern(parameters map[string]interface{}) (map[string]interface{}, error) {
	text, _ := parameters["text"].(string)
	pattern, _ := parameters["pattern"].(string)

	if pattern == "" {
		pattern = "Final Execution Complete"
	}

	re, err := regexp.Compile("(?i)" + pattern)
	if err != nil {
		return nil, err
	}

	match := re.FindString(text)

	return map[string]interface{}{
		"pattern_found": match != "",
		"pattern":       pattern,
		"matched_text":  match,
	}, nil
}

// AnalyzePerformance analyzes the cleanup agent performance
func (a *CleanupAgent) AnalyzePerformance() (map[string]interface{}, error) {
	performance := map[string]interface{}{
		"tasks_completed": a.TaskCount,
	}

	if !a.lastCleanup.IsZero() {
		performance["last_cleanup"] = a.lastCleanup.Format(time.RFC3339)
	} else {
		performance["last_cleanup"] = nil
	}

	return performance, nil
}

// SuggestImprovements suggests improvements for the cleanup agent
func (a *CleanupAgent) SuggestImprovements() ([]map[string]interface{}, error) {
	return make([]map[string]interface{}, 0), nil
}